File: Intro to Data Structures/pa5/Dictionary.c
//Galen Turoci
//gturoci
//Class CMPS 12B
//March 8th, 2018
//Dictionary.c takes a collection of (key, value) pairs and stores them,
//then implements Dictionary.h in order to allow the user to
//manipulate those pairs
//Dictionary.c

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


//private types --------------------------------------------------------------

const int tableSize = 101;

//NodeObj
typedef struct NodeObj {
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

//Node
typedef NodeObj* Node;

//newNode()
//Constructor of the Node type
Node newNode(char* k, char* v) {
	Node N = malloc(sizeof(NodeObj));
	assert(N != NULL);
	N->key = k;
	N->value = v;
	N->next = NULL;
	return(N);
}

//freeNode()
//Destructor for the Node type
void freeNode(Node* pN) {
	if (pN != NULL && *pN != NULL) {
		free(*pN);
		*pN = NULL;
	}
}

//DictionaryObj
typedef struct DictionaryObj {
	int numItems;
    Node *Table; //Hash table
} DictionaryObj;

//Public Functions------------------------------------------------------------

// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
	int sizeInBits = 8*sizeof(unsigned int);
	shift = shift & (sizeInBits - 1);
	if ( shift == 0 )
		return value;
	return (value << shift) | (value >> (sizeInBits - shift));
}

//pre_hash()
//turn a string into an unsigned int
unsigned int pre_hash(char* input) { 
	unsigned int result = 0xBAE86554;
	while (*input) {
		result ^= *input++;
		result = rotate_left(result, 5);
		}
	return result;
}
//hash()
//turns a string into an int in the range 0 to tableSize-1
int hash(char* key) {
	return pre_hash(key)%tableSize;
} 

//newDictionary()
//constructor for the Dictionary type
Dictionary newDictionary(void) {
    Dictionary D = malloc(sizeof(DictionaryObj));
    assert(D!=NULL);
    D->numItems = 0;
	D->Table = calloc(tableSize, sizeof(Node));
    return D;
}

//freeDictionary()
//destructor for the Dictionary type
void freeDictionary(Dictionary* pD) {
    if (pD!=NULL && *pD!=NULL) {
        if (!isEmpty(*pD))
            makeEmpty(*pD);
		free((*pD)->Table);
        free(*pD);
        *pD = NULL;
    }
}

//isEmpty()
//returns 1 (true) if D is empty, 0 (false) otherwise; equivalent to isEmpty
//pre: none
int isEmpty(Dictionary D) {
    if (D==NULL) {
        fprintf (stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    return(D->numItems==0);
}

//size()
//returns the amount of key, value pairs in this Dictionary
//pre: !isEmpty(D)
int size(Dictionary D) {
    if (D==NULL) {
        fprintf (stderr, "Dictionary Error: calling size() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    return D->numItems;
}

//lookup()
//returns value at position key in D
//pre: !isEmpty(D)
char* lookup(Dictionary D, char* key) {
    if (D == NULL) {
        fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
	Node N = D->Table[hash(key)];
	if (N == NULL)
		return NULL;
	else {
		while (N != NULL) {
        	if (strcmp(N->key, key) == 0)
            	return N->value;
			N = N->next;
		}
	}
	return NULL;
}

//insert()
//inserts key and value into D at location Table[hash(key)]
//pre: none
void insert(Dictionary D, char* key, char* value) {
	if (D == NULL) {
		fprintf(stderr, "Dictionary Error: calling insert() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	if (lookup(D, key) != NULL) {
		fprintf(stderr, "Dictionary Error: Cannot insert duplicate keys\n");
		exit(EXIT_FAILURE);
	}
	Node N = D->Table[hash(key)];
	if (N == NULL)
		D->Table[hash(key)] = newNode(key, value);
	else {
		N = newNode(key, value);
		N->next = D->Table[hash(key)];
		D->Table[hash(key)] = N;
	}
	D->numItems++;
}

//delete()
//deletes pair with given key from this D
//pre: !isEmpty(D)
void delete(Dictionary D, char* key) {
	if (D==NULL) {
		fprintf(stderr, "Dictionary Error: calling delete() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	Node N = D->Table[hash(key)];
	if (N == NULL) {
		fprintf(stderr, "Dictionary Error: calling delete() on non-existent key\n");
		exit(EXIT_FAILURE);
	}
	else if (strcmp(N->key, key) == 0) {
		D->Table[hash(key)] = N->next;
		freeNode(&N);
		D->numItems--;
		return;
	}
	while (strcmp(N->next->key, key) != 0)
		N = N->next;
	Node X = N->next;
	N->next = N->next->next;
	D->numItems--;
	freeNode(&X);
}

//makeEmpty()
//resets D to the empty state
//pre: !isEmpty(D)
void makeEmpty(Dictionary D) {
    if (D==NULL) {
        fprintf(stderr, "Dictionary Error: calling makeEmpty() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
	int i;
	D->numItems = 0;
	for (i = 0; i<tableSize; i++) {
		Node N = D->Table[i];
		if (N != NULL)
			D->numItems++;
	}
	i = 0;
    while (D->numItems>0) {
		Node N = D->Table[i];
		if (N != NULL) {
			Node X = N;
			D->Table[i] = D->Table[i]->next;
			freeNode(&X);
			D->numItems--;
		}
		i++;
    }
}

//printDictionary()
//prints a text representation of D to the file pointed to by out
//pre: none
void printDictionary(FILE* out, Dictionary D) {
	if (D == NULL) {
		fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	for (int i = 0; i<tableSize; i++) {
		Node N = D->Table[i];
		if (N != NULL) {
			fprintf(out, "%s %s\n", N->key, N->value);
			while (N->next != NULL) {
				N = N->next;
				fprintf(out, "%s %s\n", N->key, N->value);
			}
		}
	}
}

File: Intro to Data Structures/lab7/Dictionary.java
//Galen Turoci
//gturoci
//Class CMPS 12M
//March 7th, 2018
//Dictionary.java is a Binary Search Tree ADT that takes two strings,
//a key and a value, and places them in a Binary Search Tree.
//Dictionary.java

public class Dictionary implements DictionaryInterface {

	// Node
	private class Node {
		String key;
		String value;
		Node left;
		Node right;
		
		Node(String k, String v) {
			key = k;
			value = v;
			left = null;
			right = null;
		}
	}


	private Node head; //Reference to the first item
	private int numItems; //Number of items in the Dictionary

	//Dictionary()
	//constructor for the Dictionary class
	public Dictionary() {
		head = null;
		numItems = 0;
	}

	//findKey()
	//returns a reference to the Node at position key in this Dictionary
	//NEEDS TESTING
	private Node findKey(Node N, String key) {
		if (N == null || key.equals(N.key))
			return N;
		if (key.compareTo(N.key)<0)
			return findKey(N.left, key);
		else
			return findKey(N.right, key);
	}
	
	private Node findParent(Node N, Node M) {
		Node P = null;
		if (N != M) {
			P = M;
			while (P.right != N && P.left != N) {
				if (P.key.compareTo(N.key)<0)
					P = P.left;
				else
					P = P.right;
			}
		}
		return P;
	}
	
	private Node findLeftmost(Node N) {
		Node X = N;
		if (X != null) {
			while (N.left != null)
				N = N.left;
		}
		return N;
	}
	
	private StringBuffer printInOrder(Node N, StringBuffer s) {
		if (N != null) {
			printInOrder(N.left, s);
			s.append(N.key).append(" ");
			s.append(N.value);
			s.append("\n");
			printInOrder(N.right, s);
		}
		return s;
	}

	// public functions -----------------------------------------------------------

	// isEmpty()
	// returns true if Dictionary is empty, false otherwise
	// pre: none
	public boolean isEmpty() {
		return (numItems == 0);
	}

	// size()
	// returns the number of (key, value) pairs in this Dictionary
	// pre: none
	public int size() {
		return numItems;
	}

	// lookup()
	// returns value associated with the given key, or null reference if no such key exists
	// pre: none
	//NEEDS TESTING
	public String lookup(String key) {
		Node N = head;
		N = findKey(N, key);
		if (N == null)
			return null;
		else
			return N.value;
	}

	// insert()
	// inserts new (key,value) pair into D
	// pre: lookup(D, k)==NULL
	//NOT DONE - FIX
	public void insert(String key, String value) throws DuplicateKeyException {
		Node N = head;
		Node A, B;
		N = findKey(N, key);
		if (N != null)
				throw new DuplicateKeyException("Cannot insert duplicate keys.");
		if (isEmpty())
			head = new Node(key, value);
		else {
			N = new Node (key, value);
			B = null;
			A = head;
			while (A != null) {
				B = A;
				if (key.compareTo(A.key)<0)
					A = A.left;
				else
					A = A.right;
			}
			if (key.compareTo(B.key)<0)
				B.left = N;
			else
				B.right = N;
		}
		numItems++;
	}

	// delete()
	// deletes pair with the key k
	// pre: lookup(D, k)!=NULL
	//NEEDS TESTING
	public void delete(String key) throws KeyNotFoundException {
		Node N = head;
		Node P, S;
		N = findKey(N, key);
		if (N == null)
			throw new KeyNotFoundException("Cannot delete non-existent key.");
		if (N.left == null && N.right == null) {
			if (numItems == 1)
				head = null;
			else {
				P = findParent(N, head);
				if (P.right == N)
					P.right = null;
				else
					P.left = null;
			}
		} else if (N.right == null) {
			if (N==head)
				head = N.left;
			else {
				P = findParent(N, head);
				if (P.right == N)
					P.right = N.left;
				else
					P.left = N.left;
			}
		} else if (N.left == null) {
			if (N == head)
				head = N.right;
			else {
				P = findParent(N, head);
				if (P.right == N)
					P.right = N.right;
				else
					P.left = N.right;
			}
		} else {
			S = findLeftmost(N.right);
			N.key = S.key;
			N.value = S.value;
			P = findParent(S, N);
			if (P.right == S)
				P.right = S.right;
			else
				P.left = S.right;
		}
		numItems--;
	}

	// makeEmpty()
	// Makes this Dictionary empty
	// pre: none
	public void makeEmpty() {
		head = null;
		numItems = 0;
	}

	// printDictionary()
	// pre: none
	// prints a text representation of D to the file pointed to by out
	//NOT DONE - FIX
	public String toString() {
		StringBuffer s = new StringBuffer();
		return new String(printInOrder(head, s));
	}
}
